testcontainers/testcontainers-dotnet
# feat: Throw `DockerUnavailableException` when Docker is not available

## What does this PR do?

This pull request introduces a new `DockerUnavailableException` which is thrown instead of `ArgumentException` which was not the appropriate exception to throw since the root of the problem is not a bad argument but the state of the system which is wrong.

The exception message has also been slightly improved.

## Why is it important?

Unlike `ArgumentException`, `DockerUnavailableException` can be caught and handled.

## Related issues

Once both this pull request and #1263 are merged, the error messages will be improved, giving a better hint to the user at what went wrong. Here are two sample messages from the new `DockerUnavailableException`.

Running Testcontainers after quitting Docker Desktop:

> Docker is either not running or misconfigured. Please ensure that Docker is available at unix:///var/run/docker.sock or unix:///Users/0xced/.docker/run/docker.sock
You can customize your configuration using either the environment variables or the ~/.testcontainers.properties file. For more information, visit:
https://dotnet.testcontainers.org/custom_configuration/

Running Testcontainers after quitting [OrbStack](https://orbstack.dev):

> Docker is either not running or misconfigured. Please ensure that Docker is available at unix:///var/run/docker.sock or unix:///Users/0xced/.orbstack/run/docker.sock
You can customize your configuration using either the environment variables or the ~/.testcontainers.properties file. For more information, visit:
https://dotnet.testcontainers.org/custom_configuration/

## How to test this PR

Quit Docker Desktop or configure a bad endpoint then try to run a container. Observe the improved exception and error message.
